### PR TITLE
Positive Binary Number Convertor

### PR DESCRIPTION
**Pull request by Tanish2410**
Added a command /binary which converts a positive number to a binary number

### CODE REVIEW

**Comment by vedantgrover** [comment]
@Tanish2410 Rename the PR to be more descriptive. Idk what Binary is. Also write a description as to what you've changed in this PR and link your issue.

File: commands/util/binary.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
+const { SlashCommandBuilder, EmbedBuilder } = require("discord.js");
+
+module.exports = {
+  data: new SlashCommandBuilder()
+    .setName("binary")
+    .setDescription("Convert numbers into binary")
+    .addIntegerOption(option =>
+        option
+            .setName('integer')
```
Comment (vedantgrover) [inline commands/util/binary.js:9]: Nit: Could you rename the option to "number" or "num"? Something that's a bit more friendly than integer

File: commands/util/binary.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+            .setRequired(true)
+    ),
+  async execute(interaction) {
+
+    const integer = interaction.options.getInteger('integer')
+    const binary = integer.toString(2);
+
+    const embed = new EmbedBuilder()
+      .setTitle("Binary Converter")
+      .setDescription(`The binary representation of **${integer}** is **${binary}**.`)
```
Comment (vedantgrover) [inline commands/util/binary.js:20]: I'd say remove the title, and just add two inline fields that has the user's requested number and the binary number. The long sentence is too wordy and sometimes we just want to copy the binary and move on. Make it easier.

Personal preference is User Number on the left and Binary number is on the right.

**Review by vedantgrover — CHANGES_REQUESTED** [review]

File: commands/util/binary.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+          option
+            .setName("number")
+            .setDescription("Pick a number to convert into binary")
+            .setRequired(true)
+        )
+    )
+    .addSubcommand(subcommand =>
+      subcommand
+        .setName("add")
+        .setDescription("Add two numbers")
```
Comment (vedantgrover) [inline commands/util/binary.js:21]: Add finds the sum of two binary numbers. Not regular numbers

Also see if you can make it so that the user can add a regular number to a binary number. They can input either format in either input box

File: commands/util/binary.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      const integer = interaction.options.getInteger("integer");
+      const binary = integer.toString(2);
+
+      const embed = new EmbedBuilder()
+        .setTitle("Binary Converter")
+        .addFields(
+            {name: 'Number', value: `**${integer}**`, inline: true},
+            {name: 'Binary', value: `**${binary}**`, inline: true},
+        )
+        .setColor("BLUE");
```
Comment (vedantgrover) [inline commands/util/binary.js:48]: Remove color for now. Let's stay consistent with our color scheme until we find a color that we can use for all embeds

File: commands/util/binary.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+      const binarySum = sum.toString(2);
+
+      const embed = new EmbedBuilder()
+        .setTitle("Addition Result")
+        .addFields(
+            {name: 'First', value: `**${first}**`, inline: true},
+            {name: 'Second', value: `**${second}**`, inline: true},
+            {name: 'Sum', value: `**${binarySum}**`, inline: true},
+        )
+        .setColor("GREEN");
```
Comment (vedantgrover) [inline commands/util/binary.js:65]: Same comment as above on color scheme

File: commands/util/binary.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    ),
+  async execute(interaction) {
+    const subcommand = interaction.options.getSubcommand();
+
+    if (subcommand === "convert") {
+      const integer = interaction.options.getInteger("integer");
+      const binary = integer.toString(2);
+
+      const embed = new EmbedBuilder()
+        .setTitle("Binary Converter")
```
Comment (vedantgrover) [inline commands/util/binary.js:43]: Rename this to a success status. The user already knows it's a binary converter. You can also just remove it for a more minimalistic response

**Review by vedantgrover — CHANGES_REQUESTED** [review]

File: commands/util/binary.js (hunk 3 above)
Comment (Tanish2410) [reply commands/util/binary.js:21]: no that isnt possible 10 =/= 2

File: commands/util/binary.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
+const { SlashCommandBuilder, EmbedBuilder } = require("discord.js");
+
+module.exports = {
+  data: new SlashCommandBuilder()
+    .setName("binary")
+    .setDescription("Binary and Decimal Conversions")
+    .addSubcommand(subcommand =>
+      subcommand
+        .setName("dec_to_bin")
```
Comment (vedantgrover) [inline commands/util/binary.js:9]: This is too complicated. Change it to `encode`

File: commands/util/binary.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        .addIntegerOption(option =>
+          option
+            .setName("number")
+            .setDescription("Pick a number to convert into binary")
+            .setRequired(true)
+        )
+    )
+    .addSubcommand(subcommand =>
+        subcommand
+          .setName("bin_to_dec")
```
Comment (vedantgrover) [inline commands/util/binary.js:20]: Change it to `decode`

File: commands/util/binary.js [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        const first = interaction.options.getInteger("first");
+        const second = interaction.options.getInteger("second");
+        const sum = parseInt(first, 2) + parseInt(second, 2);
+        const binarySum = sum.toString(2);
+  
+        const embed = new EmbedBuilder()
+            .addFields(
+                {name: 'First', value: `**${first}**`, inline: true},
+                {name: 'Second', value: `**${second}**`, inline: true},
+                {name: 'Sum', value: `**${binarySum}**`, inline: true},
```
Comment (vedantgrover) [inline commands/util/binary.js:82]: Okay this is good but can you make it one field with an expression?

Like first + second?

**Review by vedantgrover — CHANGES_REQUESTED** [review]

File: commands/util/binary.js [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+          .setName("decode")
+          .setDescription("Convert a binary number to a decimal number")
+          .addIntegerOption(option =>
+            option
+              .setName("binary")
+              .setDescription("Pick a binary number to convert into decimals")
+              .setRequired(true)
+        )
+    )
+    .addSubcommand(subcommand =>
```
Comment (vedantgrover) [inline commands/util/binary.js:29]: A remark in general:
Program your arithmetic functions in such a way that it doesn't matter what the user puts in, the bot does the arithmetic. 

If the user put in a regular integer, it converts it to binary and does the math. If the user puts in a binary number, it keeps it as a binary number and does the arithmetic

**Review by vedantgrover — CHANGES_REQUESTED** [review]

File: commands/util/binary.js (hunk 10 above)
Comment (Tanish2410) [reply commands/util/binary.js:29]: bro you can't do that 10 is a decimal and a binary number

File: commands/util/binary.js [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+          await interaction.reply({ embeds: [embed] });
+        } catch (error) {
+          await interaction.reply({ content: `An error occurred while decoding: ${error.message}`, ephemeral: true });
+        }
+
+      } else if (subcommand === "add") {
+        try {
+          const first = interaction.options.getInteger("first");
+          const second = interaction.options.getInteger("second");
+          if (!/^[01]+$/.test(first.toString()) || !/^[01]+$/.test(second.toString())) {
```
Comment (vedantgrover) [inline commands/util/binary.js:124]: Not a problem right now but later on down the line, please separate the boolean and make it its own variable for better code readability

File: commands/util/binary.js [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+            throw new Error("Both inputs must be valid binary numbers.");
+          }
+          const sum = parseInt(first, 2) + parseInt(second, 2);
+          const binarySum = sum.toString(2);
+
+          const embed = new EmbedBuilder()
+            .addFields(
+              { name: 'First + Second', value: `First: **${first}**, Second: **${second}**`, inline: true },
```
Comment (vedantgrover) [inline commands/util/binary.js:132]: Your value should be `first + second` and your field title should be `Expression`

File: commands/util/binary.js [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+          const second = interaction.options.getInteger("second");
+          if (!/^[01]+$/.test(first.toString()) || !/^[01]+$/.test(second.toString())) {
+            throw new Error("Both inputs must be valid binary numbers.");
+          }
+          const difference = parseInt(first, 2) - parseInt(second, 2);
+          const binaryDifference = difference.toString(2);
+
+          const embed = new EmbedBuilder()
+            .addFields(
+              { name: 'First - Second', value: `First: **${first}**, Second: **${second}**`, inline: true },
```
Comment (vedantgrover) [inline commands/util/binary.js:153]: Same as above except with `-`

File: commands/util/binary.js [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+          const second = interaction.options.getInteger("second");
+          if (!/^[01]+$/.test(first.toString()) || !/^[01]+$/.test(second.toString())) {
+            throw new Error("Both inputs must be valid binary numbers.");
+          }
+          const product = parseInt(first, 2) * parseInt(second, 2);
+          const binaryProduct = product.toString(2);
+
+          const embed = new EmbedBuilder()
+            .addFields(
+              { name: 'First * Second', value: `First: **${first}**, Second: **${second}**`, inline: true },
```
Comment (vedantgrover) [inline commands/util/binary.js:174]: Same as above except with `*`

File: commands/util/binary.js (hunk 12 above)
Comment (vedantgrover) [reply commands/util/binary.js:132]: Make sure that it actually shows the expression. If the user gives first number as 01 and the second number as 10, it should show 01 + 10

**Review by vedantgrover — CHANGES_REQUESTED** [review]

**Comment by vedantgrover** [comment]
Also @Tanish2410 are you gonna do division?

File: commands/util/binary.js [hunk 15]
```diff
@@ -0,0 +1,191 @@
...
+      } else if (subcommand === "subtract") {
+        try {
+          const first = interaction.options.getInteger("first");
+          const second = interaction.options.getInteger("second");
+          const check = !/^[01]+$/;
+          if (check.test(first.toString()) || check.test(second.toString())) {
+            throw new Error("Both inputs must be valid binary numbers.");
+          }
+          const difference = parseInt(first, 2) - parseInt(second, 2);
+          const binaryDifference = difference.toString(2);
```
Comment (vedantgrover) [inline commands/util/binary.js:152]: What happens if the binaryDifference is negative? Does it still work? Not sure but was wondering if you tested it out.

Comment (Tanish2410) [reply commands/util/binary.js:152]: negative numbers don't work rn just positive numbers

Comment (vedantgrover) [reply commands/util/binary.js]: Yeah but what does it give you? As the output. There aren't any checks in place theree

Comment (Tanish2410) [reply commands/util/binary.js]: no it works

Comment (Tanish2410) [reply commands/util/binary.js]: i just tested it

Comment (Tanish2410) [reply commands/util/binary.js]: I went to a website and the negative version of a binary number is just -and then the number in binary

File: commands/util/binary.js [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+          throw new Error("Both inputs must be valid binary numbers.");
+        }
+
+        const sum = parseInt(first, 2) + parseInt(second, 2);
+        const binarySum = sum.toString(2);
+
+        const embed = new EmbedBuilder()
+          .addFields(
+            { name: "First + Second", value: `First: **${first}** + Second: **${second}**`, inline: true },
+            { name: "Sum", value: `**${binarySum}**`, inline: true }
```
Comment (vedantgrover) [inline commands/util/binary.js:145]: Just give me the number. `01 + 10101`. No labels. I don't want to know which number is first and which number is second. Change the field title to be `Expression`. Not `First + Second`.

Please do this for all subcommands

**Review by vedantgrover — CHANGES_REQUESTED** [review]

File: commands/util/binary.js (hunk 16 above)
Comment (Tanish2410) [reply commands/util/binary.js:145]: why?

Comment (Tanish2410) [reply commands/util/binary.js:145]: what's so bad about being more clear and specific?

Comment (GavinDPinto) [reply commands/util/binary.js:145]: It looks cleaner; also its implied what operations are being performed.

Comment (vedantgrover) [reply commands/util/binary.js:145]: Having First and Second as labels is unnecessary. The users knows what numbers they put in and they don't need that stated to them at all times. Also the embed is getting overly complicated and overwhelming to look at with so much text.

Another recommendation I have is don't make it inline. That way the embed has room to expand horizontally and won't interfere with the other field.

File: commands/util/binary.js [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+            { name: "Expression", value: `${integer}`, inline: true },
+            { name: "Result", value: `${binary}`, inline: true }
+          );
+
+        await interaction.reply({ embeds: [embed] });
+
+      } else if (subcommand === "decode") {
+        const binary = interaction.options.getInteger("binary").toString();
+        if (!validateBinary(binary)) {
+          throw new Error("Input must be a valid binary number.");
```
Comment (vedantgrover) [inline commands/util/binary.js:119]: Don't throw an error because that's going to crash the bot.

Send an ephemeral message that lets the user know that the input has to be a valid binary number.

File: commands/util/binary.js [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+
+        await interaction.reply({ embeds: [embed] });
+
+      } else if (subcommand === "add") {
+        const first = interaction.options.getInteger("first").toString();
+        const second = interaction.options.getInteger("second").toString();
+
+        if (!validateBinary(first) || !validateBinary(second)) {
+          throw new Error("Both inputs must be valid binary numbers.");
+        }
```
Comment (vedantgrover) [inline commands/util/binary.js:137]: Same thing as above

File: commands/util/binary.js [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+
+        await interaction.reply({ embeds: [embed] });
+
+      } else if (subcommand === "subtract") {
+        const first = interaction.options.getInteger("first").toString();
+        const second = interaction.options.getInteger("second").toString();
+
+        if (!validateBinary(first) || !validateBinary(second)) {
+          throw new Error("Both inputs must be valid binary numbers.");
+        }
```
Comment (vedantgrover) [inline commands/util/binary.js:156]: Same thing

File: commands/util/binary.js [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+
+        await interaction.reply({ embeds: [embed] });
+
+      } else if (subcommand === "multiply") {
+        const first = interaction.options.getInteger("first").toString();
+        const second = interaction.options.getInteger("second").toString();
+
+        if (!validateBinary(first) || !validateBinary(second)) {
+          throw new Error("Both inputs must be valid binary numbers.");
+        }
```
Comment (vedantgrover) [inline commands/util/binary.js:175]: Same thing

File: commands/util/binary.js [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+          option
+            .setName("second")
+            .setDescription("The second number")
+            .setRequired(true)
+        )
+    ),
+
+  async execute(interaction) {
+    try {
+      const subcommand = interaction.options.getSubcommand();
```
Comment (vedantgrover) [inline commands/util/binary.js:100]: You don't really need a try catch because this isn't super high stakes. I'd say just send an error message to the user if something goes wrong with the user input.

File: commands/util/binary.js [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+
+        await interaction.reply({ embeds: [embed] });
+
+      } else if (subcommand === "divide") {
+        const first = interaction.options.getInteger("first").toString();
+        const second = interaction.options.getInteger("second").toString();
+
+        if (!validateBinary(first) || !validateBinary(second)) {
+          throw new Error("Both inputs must be valid binary numbers.");
+        }
```
Comment (vedantgrover) [inline commands/util/binary.js:194]: Same thing

File: commands/util/binary.js [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+
+        if (parseInt(second, 2) === 0) {
+          throw new Error("Division by zero is not allowed.");
+        }
```
Comment (vedantgrover) [inline commands/util/binary.js:198]: Again send a message to the user. They don't know what went wrong

File: commands/util/binary.js [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+          .addFields(
+            { name: "Expression", value: `${first} ÷ ${second}`, inline: true },
+            { name: "Result", value: `${binaryQuotient}`, inline: true }
+          );
+
+        await interaction.reply({ embeds: [embed] });
+      }
+    } catch (error) {
+      await interaction.reply({ content: `An error occurred: ${error.message}`, ephemeral: true });
+    }
```
Comment (vedantgrover) [inline commands/util/binary.js:213]: While this does work, it's messy and will display ANY error to the user which we don't want. Just send the user the information they need to care about.

Also see if you can make it into an embed with some error markings (either the color or there's an X to indicate something wrong)

File: commands/util/binary.js (hunk 17 above)
Comment (vedantgrover) [reply commands/util/binary.js:119]: Also here you don't need to tell the user that an error occurred. Make sure the embed styling sends that message. Just have the text be "Input must be a valid binary number."

**Review by vedantgrover — APPROVED** [review]
Type shit
